### PR TITLE
Mqtt username password authentication

### PR DESCRIPTION
**Pull request by lprashant-94**

### CODE REVIEW

File: altizon/io/datonis/edge/edge_gateway_mqtt.py [hunk 1]
```diff
@@ -16,13 +17,33 @@
...
+def on_connect(client, userdata, flags, rc):
+    if rc == mqtt.CONNACK_ACCEPTED:
+        userdata.state = CONNECTED
+        logging.info("Connected to the MQTT broker with return code: " + str(flags) + ", " + str(rc))
+        userdata.subscribe_for_instructions()
+        userdata.subscribe_for_acks()
+    elif rc == mqtt.CONNACK_REFUSED_NOT_AUTHORIZED:
+        logging.error("Connection Unauthorised. ")
+        userdata.state = UNAUTHORISED
+        client.disconnect()
```
Comment (rajesh-altizon) [inline altizon/io/datonis/edge/edge_gateway_mqtt.py]: Do you still need to explicitly disconnect? Or will disconnect raise an exception indicating already not connected?

File: altizon/io/datonis/edge/edge_gateway_mqtt.py [hunk 2]
```diff
@@ -16,13 +17,33 @@
...
+    else:
+        userdata.state = RECONNECTING
 
 
 def on_disconnect(client, userdata,rc):
+    if rc == mqtt.MQTT_ERR_SUCCESS and userdata.state != UNAUTHORISED:
+        userdata.state = DISCONNECTED
+    elif userdata.state != UNAUTHORISED:
+        userdata.state = RECONNECTING
```
Comment (rajesh-altizon) [inline altizon/io/datonis/edge/edge_gateway_mqtt.py]: Better way:

````
if userdata.state != UNAUTHORISED:
    if rc == mqtt.MQTT_ERR_SUCCESS:
         userdata.state = DISCONNECTED
    else:
         userdata.state = RECONNECTING

File: altizon/io/datonis/edge/edge_gateway_mqtt.py [hunk 3]
```diff
@@ -45,8 +66,7 @@ def instruction_worker(thread_name,gateway):
         try:
             instruction_dispatcher(gateway) 
         except:
-            e = sys.exc_info()[0]
-            logging.error('instruction_dispatcher failed', str(e))
+            logging.error('instruction_dispatcher failed', exc_info=True)
```
Comment (rajesh-altizon) [inline altizon/io/datonis/edge/edge_gateway_mqtt.py]: Did not understand what happens here. Does it log the full exception?

File: altizon/io/datonis/edge/edge_gateway_mqtt.py [hunk 4]
```diff
@@ -102,10 +126,9 @@ def connect(self):
         if retval == 0:
             self.ack_lock = threading.Condition()
             self.mqtt_client.loop_start()
+            self.state = CONNECTING
```
Comment (rajesh-altizon) [inline altizon/io/datonis/edge/edge_gateway_mqtt.py]: should be CONNECTED right?

File: altizon/io/datonis/edge/edge_gateway_mqtt.py [hunk 5]
```diff
@@ -102,10 +126,9 @@ def connect(self):
...
             # Start a new thread for instruction execution
             thread.start_new_thread(instruction_worker, ('instruction-worker', self))
-            self.subscribe_for_instructions()
```
Comment (rajesh-altizon) [inline altizon/io/datonis/edge/edge_gateway_mqtt.py]: Were we subscribing for instructions/acks twice? Once here and once in on_connect()?

File: altizon/io/datonis/edge/edge_gateway_mqtt.py (hunk 3 above)
Comment (lprashant-94) [reply altizon/io/datonis/edge/edge_gateway_mqtt.py]: Yes, logging module print complete stack trace if exc_info parameter is passed.

File: altizon/io/datonis/edge/edge_gateway_mqtt.py (hunk 5 above)
Comment (lprashant-94) [reply altizon/io/datonis/edge/edge_gateway_mqtt.py]: Yes, on_connect callback subscribes for instruction after connection.

**Comment by lprashant-94** [comment]
@rajesh-altizon  I have made all suggested changes, Please review.

**Pull request merged**